src/Microsoft.Framework.DesignTimeHost/ProtocolManager.cs

```diff
@@ -0,0 +1,149 @@
+
+            if (_environmentOverriden)
+            {
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: DTH protocol negotiation is override by environment variable {EnvDthProtocol} and set to {CurrentVersion}.");
+                return;
+            }
+
+            if (message.MessageType == NegotiationMessageTypeName)
+            {
+                if (message.ContextId != 0)
+                {
+                    return;
+                }
+
+                InternalNegotiate(message, needRespond: true);
+            }
+            else if (message.MessageType == InitializeMessageTypeName)
+            {
+                InternalNegotiate(message, needRespond: false);
+            }
+        }
+
+        private void InternalNegotiate(Message message, bool needRespond = false)
+        {
+            var tokenValue = message.Payload?["Version"];
+            if (tokenValue == null)
+            {
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication failed. Version property is missing in payload.");
+                return;
+            }
+
+            var preferredVersion = tokenValue.Value<int>();
+            if (preferredVersion == 0)
+            {
+                // the preferred version can't be zero. either property is missing or the the payload is corrupted.
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication failed. Protocol version 0 is invalid.");
+                return;
+            }
+
+            int? protocol = null;
+            for (int i = 0; i < SupportedProtocolVersions.Length; ++i)
+            {
+                if (SupportedProtocolVersions[i] <= preferredVersion)
+                {
+                    protocol = SupportedProtocolVersions[i];
+                    break;
+                }
+            }
+
+            if (protocol.HasValue)
+            {
+                CurrentVersion = protocol.Value;
+                NeedToRespond = needRespond;
+
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication successed. Use protocol {CurrentVersion}");
+            }
+            else
+            {
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication failed. Stay on protocol {CurrentVersion}");
+            }
+        }
+
+        public void Respond(ConnectionContext context)
+        {
+            Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Respond to protocol negotiation.");
+
+            context.Transmit(new Message
+            {
+                ContextId = 0,
+                MessageType = NegotiationMessageTypeName,
+                Payload = JToken.FromObject(new { Version = CurrentVersion })
+            });
+            
+            NeedToRespond = false;
+        }
+
+        private static int? GetProtocolVersionFromEnvironment()
```

Comment:
Allows the environment variable overridden is useful when developing new protocol.